bqskit/runtime/detached.py

```diff
@@ -0,0 +1,571 @@
+from bqskit.runtime.task import RuntimeTask
+
+
+def listen(server: DetachedServer, listener: Listener) -> None:
+    """Listening thread listens for client connections."""
+    while server.running:
+        client = listener.accept()
+        server.logger.debug('Connected to client.')
+        if server.running:
+            server.clients[client] = set()
+            server.sel.register(client, selectors.EVENT_READ, 'from_client')
+            server.logger.info('Registered client.')
+
+
+def send_outgoing(server: DetachedServer) -> None:
+    """Outgoing thread forwards messages as they are created."""
+    while server.running:
+        if len(server.outgoing) > 0:
+            time.sleep(0)
+            outgoing = server.outgoing.pop()
+            outgoing[0].send((outgoing[1], outgoing[2]))
+
+            server.logger.debug(f'Sent message {outgoing[1]}.')
+            server.logger.log(1, f'{outgoing[2]}\n')
+
+
+class DetachedServer:
+    """
+    BQSKit Runtime Server in detached mode.
+
+    In detached mode, the runtime is started separately from the client. Clients
+    can connect and disconnect rather than shutdown a detached server. This
+    architecture is designed for the distributed setting, where managers manage
+    workers in shared memory and communicate with the server over a network.
+    """
+
+    def __init__(
+        self,
+        ipports: list[tuple[str, int]],
+        port: int = 7472,
+    ) -> None:
+        """
+        Create a server and connect to the managers at `ipports`.
+
+        Args:
+            ipports (list[tuple[str, int]]): The ip and port pairs were
+                managers are expected to be waiting.
+
+            port (int): The port this server will listen for clients on.
+                Default is 7472.
+        """
+
+        self.clients: dict[Connection, set[uuid.UUID]] = {}
+        """Tracks all connected clients and all the tasks they have created."""
+
+        self.tasks: dict[uuid.UUID, tuple[int, Connection]] = {}
+        """Tracks all active CompilationTasks submitted to the cluster."""
+
+        self.mailbox_to_task_dict: dict[int, uuid.UUID] = {}
+        """Used to convert internal RuntimeTasks to client CompilationTasks."""
+
+        self.mailboxes: dict[int, Any] = {}
+        """A server mailbox is where CompilationTask results are stored."""
+
+        self.mailbox_counter = 0
+        """Counter to ensure all mailboxes have unique IDs."""
+
+        self.managers: list[Connection] = []
+        """Maintains one duplex connection with all connected managers."""
+
+        self.manager_resources: list[int] = []
+        """Tracks the total number of workers each manager manages."""
+
+        self.sel = selectors.DefaultSelector()
+        """Used to efficiently idle and wake when communication is ready."""
+
+        self.running = True
+        """True while the server is running."""
+
+        self.logger = logging.getLogger('bqskit-runtime')
```

Comment:
The detached and manager loggers are explicitly separate from the rest of the logging system. This is because these loggers should not have their log records passed around in the system. Rather, they should be handled in their respective processes.